server/src/main/java/org/elasticsearch/snapshots/InternalSnapshotsInfoService.java

```diff
@@ -132,6 +142,7 @@ public void clusterChanged(ClusterChangedEvent event) {
                     if (knownSnapshotShardSizes.containsKey(snapshotShard) == false) {
                         // check if already fetching snapshot info in progress
                         if (unknownSnapshotShards.add(snapshotShard)) {
+                            failedSnapshotShards.remove(snapshotShard); // retry the failed shard
```

Comment:
I think we should also remove this in `cleanUpKnownSnapshotShardSizes` to avoid this eventually potentially leading to a memory issue. We can do this in a follow-up too, no need to hold merging this PR.

Reply:
I opened https://github.com/elastic/elasticsearch/pull/63492 to address this